app/controllers/clock_controller.rb

```diff
@@ -1,14 +1,24 @@
 class ClockController < ApplicationController
+  def main
+    current_timezone
+    @current_weather = current_weather(@current_timezone.name)
+  end
+
   def index
-    @current_timezone = Time.zone
+    current_timezone
+    @current_weather = current_weather(@current_timezone.name)
+    @user_timezones = current_user.timezones
   end
 
   def timezones
-    @timezones = ActiveSupport::TimeZone.all
+    @timezones = user_signed_in? ? all_timezones[1..5].map do |timezone|
+      {timezone: timezone, weather: current_weather(timezone.name)}
+    end : all_timezones.map { |timezone| {timezone: timezone} }
```

Comment:
What do you think about handling this with a service object?

Reply:
something like this: 

```suggestion

    @timezones = user_signed_in? ? WeatherSetter(all_timezones.first(5)) :
      all_timezones

class WeatherSetter
  def initialize(timezones)
    @timezones = timezones
  end

  def call
    set_timezones_with_weather
  end

  def set_timezones_with_weather
    @timezones.map do |timezone|
      class << timezone
        def weather
          CurrentWeather.call(timezone.name)
        end
      end
    end
  end
end
```

Reply:
also, another option could be using an OpenStruct instead of adding the method to each instance: 
```suggestion
 def set_timezones_with_weather
    @timezones.map do |timezone|
      OpenStruct.new(
        name:      timezone.name,
        weather:  CurrentWeather.call(timezone.name))
    end
  end
```